src/core/a-assets.js

```diff
@@ -10,126 +10,129 @@ var warn = debug('core:a-assets:warn');
 /**
  * Asset management system. Handles blocking on asset loading.
  */
-module.exports = registerElement('a-assets', {
-  prototype: Object.create(ANode.prototype, {
-    createdCallback: {
-      value: function () {
-        this.isAssets = true;
-        this.fileLoader = fileLoader;
-        this.timeout = null;
-      }
-    },
-
-    attachedCallback: {
-      value: function () {
-        var self = this;
-        var i;
-        var loaded = [];
-        var mediaEl;
-        var mediaEls;
-        var imgEl;
-        var imgEls;
-        var timeout;
-
-        if (!this.parentNode.isScene) {
-          throw new Error('<a-assets> must be a child of a <a-scene>.');
-        }
+class AAssets extends ANode {
+  constructor () {
+    super();
+    this.isAssets = true;
+    this.fileLoader = fileLoader;
+    this.timeout = null;
+  }
 
-        // Wait for <img>s.
-        imgEls = this.querySelectorAll('img');
-        for (i = 0; i < imgEls.length; i++) {
-          imgEl = fixUpMediaElement(imgEls[i]);
-          loaded.push(new Promise(function (resolve, reject) {
-            // Set in cache because we won't be needing to call three.js loader if we have.
-            // a loaded media element.
-            THREE.Cache.add(imgEls[i].getAttribute('src'), imgEl);
-            imgEl.onload = resolve;
-            imgEl.onerror = reject;
-          }));
-        }
+  connectedCallback () {
+    // Defer if DOM is not ready.
+    if (document.readyState === 'loading') {
+      document.addEventListener('DOMContentLoaded', this.connectedCallback.bind(this));
```

Comment:
It's effectively the same but this way when read it's more clear that we're just deferring the call. Don't have to read understand `doConnectedCallback`

Reply:
evaluating the condition only happens once so perf-wise is negligible over more clear code

Reply:
In your logic, if it's more clear to you to see that we're just deferring the call, and I can agree with that, then you don't even need to create a separate function `doConnectedCallback`, you can just put the code inside `connectedCallback` after the condition, unless I'm missing something why you created it. That's why it bothered me reading this code, when I have this pattern I usually create a function like this to use it as the callback, mostly important when the condition is costly so it's not executed a second time. Here I agree the condition cost is negligible.

To me, the condition is evaluated twice when the readyState is loading:

- `connectedCallback` called
- `if (document.readyState === 'loading')` evaluated to true
-  `document.addEventListener('DOMContentLoaded', this.connectedCallback.bind(this));` callback registered
- `connectedCallback` called because of the DOMContentLoaded event
- `if (document.readyState === 'loading')` evaluated to false
- `doConnectedCallback` called

in comparison, using `doConnectedCallback` as the callback:

- `connectedCallback` called
- `if (document.readyState === 'loading')` evaluated to true
-  `document.addEventListener('DOMContentLoaded', this.doConnectedCallback.bind(this));` callback registered
- `doConnectedCallback` called because of the DOMContentLoaded event

Reply:
I created it for readability. I like small focused functions